lib/node/runner/api_message_handler_test.go

```diff
@@ -0,0 +1,200 @@
+package runner
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	"github.com/gorilla/mux"
+	"github.com/stellar/go/keypair"
+	"github.com/stretchr/testify/require"
+
+	"boscoin.io/sebak/lib/common"
+	"boscoin.io/sebak/lib/error"
+	"boscoin.io/sebak/lib/network"
+	"boscoin.io/sebak/lib/network/api"
+	"boscoin.io/sebak/lib/network/httputils"
+	"boscoin.io/sebak/lib/transaction"
+	"boscoin.io/sebak/lib/transaction/operation"
+)
+
+type HelperTestNodeMessageHandler struct {
+	HelperTestGetNodeTransactionsHandler
+
+	conf        common.Config
+	nodeHandler *NetworkHandlerNode
+}
+
+func (p *HelperTestNodeMessageHandler) Prepare() {
+	p.HelperTestGetNodeTransactionsHandler.Prepare()
+
+	p.conf = common.NewConfig()
+	p.nodeHandler = NewNetworkHandlerNode(
+		p.localNode,
+		p.network,
+		p.st,
+		p.consensus,
+		network.UrlPathPrefixNode,
+		p.conf,
+	)
+
+	// override existing handler
+	p.router = mux.NewRouter()
+	p.server = httptest.NewServer(p.router)
+	p.router.HandleFunc(api.PostTransactionPattern, p.nodeHandler.MessageHandler).
+		Methods("POST").
+		MatcherFunc(common.PostAndJSONMatcher)
+}
+
+func (p *HelperTestNodeMessageHandler) URL(urlValues url.Values) (u *url.URL) {
+	u, _ = url.Parse(p.server.URL)
+	u.Path = api.PostTransactionPattern
+
+	if urlValues != nil {
+		u.RawQuery = urlValues.Encode()
+	}
+
+	return
+}
+
+func (p *HelperTestNodeMessageHandler) makeTransaction() (tx transaction.Transaction) {
+	receiverKP, _ := keypair.Random()
+	tx = transaction.MakeTransactionCreateAccount(p.genesisKeypair, receiverKP.Address(), common.BaseReserve)
+	tx.B.SequenceID = p.genesisAccount.SequenceID
+	tx.Sign(p.genesisKeypair, networkID)
+
+	return
+}
+
+func TestNodeMessageHandler(t *testing.T) {
+	p := &HelperTestNodeMessageHandler{}
+	p.Prepare()
+	defer p.Done()
+
+	u := p.URL(nil)
+
+	tx := p.makeTransaction()
+	require.Nil(t, tx.IsWellFormed(networkID, p.conf))
+
+	postData, _ := tx.Serialize()
+	req, err := http.NewRequest("POST", u.String(), bytes.NewBuffer(postData))
+	req.Header.Set("Content-Type", "application/json")
+	require.Nil(t, err)
+	resp, err := p.server.Client().Do(req)
+	require.Nil(t, err)
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+	require.True(t, p.consensus.TransactionPool.Has(tx.GetHash()))
+}
+
+func TestNodeMessageHandlerNotWellformedTransaction(t *testing.T) {
+	p := &HelperTestNodeMessageHandler{}
+	p.Prepare()
+	defer p.Done()
+
+	u := p.URL(nil)
+
+	{ // invalid signature
+		tx := p.makeTransaction()
+		tx.H.Signature = "findme"
+		errIsWellformed := tx.IsWellFormed(networkID, p.conf)
+		require.Equal(t, errors.ErrorInvalidTransaction.Code, errIsWellformed.(*errors.Error).Code)
```

Comment:
Why not just compare the errors ? Avoids the cast

Reply:
```
lib/node/runner/api_message_handler_test.go:104:72: errIsWellformed.Code undefined (type error has no field or method Code)
```
:)

Reply:
`require.Equal(t, errors.ErrorInvalidTransaction, errIsWellformed)`

:)

Reply:
@Geod24 :) :)

```
=== RUN   TestNodeMessageHandlerNotWellformedTransaction
--- FAIL: TestNodeMessageHandlerNotWellformedTransaction (0.01s)
    require.go:157:
        	Error Trace:	api_message_handler_test.go:104
        	Error:      	Not equal:
        	            	expected: &errors.Error{Code:0xa4, Message:"invalid transaction", Data:map[string]interface {}{}}
        	            	actual  : &errors.Error{Code:0xa4, Message:"invalid transaction", Data:map[string]interface {}{"error":"signature verification failed"}}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(*errors.Error)({"code":164,"message":"invalid transaction","data":{}})
        	            	+(*errors.Error)({"code":164,"message":"invalid transaction","data":{"error":"signature verification failed"}})

        	Test:       	TestNodeMessageHandlerNotWellformedTransaction
FAIL
```

Reply:
I see.

Then could we have an helper function somewhere which compare those errors, with documentation on why it's useful. Looking at this code, I cannot tell that the returned error has extra information. And I've seen this pattern used in a few places already.